### DISCUSSION TITLE
Gracefully power off adapter

### DISCUSSION THREAD

**Discussion by manio** [bluez/bluer#157] [Q&A] [answered]
Hi,
I want you to ask how to properly power down bluetooth adapter.
My app is starting session, creating bluetooth adapter, doing some tasks and finally powers it down (dropping the session, adapter and all handlers as it goes out of the scope), after some time it is doing it over again.
I saw in examples that the BLE Advertisements and Profile register are intentionally dropped with `drop()`.

I thought that I could just do:
```
adapter.set_powered(false);
```
and then just leave rust make it cleanup work but I am not sure if it is OK.

Do I need to drop() all handlers before powering down?
I also noticed this:
```
bluetoothd[164]: src/adv_monitor.c:btd_adv_monitor_power_down() Unexpected NULL btd_adv_monitor_manager object upon power down
```
which leads me to think that powering the adapter down and let rust do its work maybe is not OK?

**Comment by surban** [comment, marked as answer]
BlueR itself is not doing any cleanup when you power down the adapter. It just sends the power down request to `bluetoothd`.

I am not aware of any explicit requirements to remove advertisements before powering down the adapter or quitting the program. In our examples the explicit `drop` is there to test that BlueR properly unregisters dropped objects. However, as a system service `bluetoothd` is expected to clean up after programs that just exit.

If your code is working fine, I would not worry about the cleanup sequence.
